### ISSUE TITLE
NN product training runtime: state, optimizers, RNG, checkpoints, and numerical policy

### ISSUE THREAD

**Issue by iteathen** [iteathen/CUDA-JS#80]
## Audit reconciliation — 2026-09-01

Retained as future **separate NN publish-unit** semantics. This issue must not put optimizer/model/training state into the `cuda-js` core package.

## Current state

```text
parent:                #70 separate NN publish unit
dependencies:          #72, #73, #74 plus accepted execution/provider/memory plans
implementation status: not implemented
priority:              deferred
```

## Owned outcome when dependency-ready

Own the explicit logical training state machine: parameter/state registry, deterministic initialization/RNG, loss declaration, gradient accumulation/reset, optimizer transitions, numerical policy, loss scaling where selected, consistent step outcomes, checkpoints/restore, bounded metrics and session lifecycle.

Native memory, transfers, prepared execution, provider handles and cleanup remain public CUDA-JS core ownership. Generic Tensor math/spec semantics are consumed from their natural owner.

## First profile

Plain deterministic FP32 SGD must be fully specified and independently qualified before AdamW, mixed precision, clipping, dropout, dynamic loss scaling or distributed state.

A failed/unproved device update must never be reported as an intact old or clean new training state; restore/restart requirements must remain explicit.

## Acceptance before implementation

- [ ] Upstream tensor/graph/autodiff/package boundaries are accepted.
- [ ] State/RNG/optimizer/checkpoint identities and transitions are deterministic and provider-neutral.
- [ ] Independent CPU references cover initialization, SGD and later admitted optimizers.
- [ ] Checkpoint continuation matches uninterrupted execution within the declared profile.
- [ ] No core CUDA-JS API learns NN/model/optimizer semantics.

## Non-goals

No same-package `cuda-js/nn`, arbitrary mutation of live parameters, every optimizer in v1, higher-order differentiation, hidden CPU fallback, native provider state in checkpoints, or implementation authority from this issue alone.
